Tests/BasicTests/GraphAlgorithmsTests.swift

```diff
@@ -70,8 +74,36 @@ class GraphAlgorithmsTests: XCTestCase {
         XCTAssertThrows(GraphError.unexpectedCycle) { _ = try topologicalSort(1, [1: [2], 2: [1]]) }
     }
 
+    func testCycleDetection() throws {
+        // Single node graph.
+        XCTAssertNotCycle(findCycle(1, [:]))
+        XCTAssertNotCycle(findCycle(1, [1: [2]]))
+        // Trivial cycles.
+        XCTAssertCycle(findCycle(1, [1: [1]]), [1])
+        XCTAssertCycle(findCycle(1, [1: [2], 2: [1]]), [1, 2])
+        XCTAssertCycle(findCycle(1, [1: [2], 2: [3], 3: [2]]), [2, 3])
+        XCTAssertCycle(findCycle(1, [1: [2], 2: [3], 3: [1]]), [1, 2, 3])
+
+        XCTAssertNotCycle(findCycle(1, [1: [2, 3], 2: [3, 4], 3: [4, 5], 4: [5, 8], 5: [7, 8], 7: [8]]))
+        XCTAssertCycle(findCycle(1, [1: [2], 2: [3, 4], 3: [4, 5], 4: [1, 5, 8], 5: [7, 8], 7: [8]]), [1, 2, 3, 4])
+    }
+
```

Comment:
There should be a test of this situation:
```
1: [2, 3]
2: []
3: [2]
```
which checks the stack is popped (I think these tests will all pass if you remove the pop in the implementation).